teps/0026-interceptor-plugins.md

```diff
@@ -0,0 +1,510 @@
+
+
+## Design Details
+
+This section describes the two major portions of the proposal -- the changes to
+the Interceptor interface to make it more typed, and the new CRD to
+install/create interceptors.
+
+### HTTP Interface
+
+This diverges from the current HTTP interface that simply forwards the given request and uses HTTP response codes. Instead the proposed way uses typed response bodies for both requests and response (similar to the AdmissionReviewRequest and Response in AdmissionControllers)
+
+
+#### Request
+
+EL Sink sends incoming events to interceptors over HTTP using a typed format:
+
+```go
+type InterceptorRequest struct {  
+	// Body is the incoming HTTP event body
+	Body []byte `json:"body,omitempty"`
+	// Header are the headers for the incoming HTTP event
+	Header map[string][]string `json:"header,omitempty"`
+
+	// InterceptorParams are the user specified params for the interceptor
+	InterceptorParams map[string]string `json:"interceptor_params,omitempty"`
+
+	Context *TriggerContext
+}
+
+type TriggerContext struct {
+	// EventUrl is the URL of the incoming event
+	EventUrl string `json:"url,omitempty"`
+	// EventID is a unique ID assigned by Triggers to each event
+	EventID string `json:"event_id,omitempty"`
+	// TriggerID is of the form namespace/$ns/triggers/$name
+	TriggerID string `json:"trigger_id,omitempty"`
+}
+```
+
+#### Response
+
+Interceptors in turn return a message with a fixed type:
+
+```go
+type InterceptorResponse struct {
+	// Extensions are  additional fields that is added to the interceptor event.
+	// See TEP-0022. Naming TBD.
+	Extensions map[string]interface{} `json:"extensions,omitempty"`
+	// Continue indicates if the EventListener should continue processing the Trigger or not
+	Continue bool `json:"continue,omitempty"`
+	// Status is an Error status t
+	Status grpc.Status `json:"status,omitempty"`
+}
+```
+
+#### Versioning
+
+To allow changes to the Interceptor HTTP interface, we can version the request and response types by embedding them in a `InterceptorMessage` type which contains an inline `TypeMeta`:
+
+```go
+type InterceptorMessage struct {
+    metav1.TypeMeta `json:",inline"`
+    // +optional
+    Request *InterceptorRequest `json:"request,omitempty" protobuf:"bytes,1,opt,name=request"`
+    // +optional
+    Response *InterceptorResponse `json:"response,omitempty" protobuf:"bytes,2,opt,name=response"`
```

Comment:
im a bit confused about the mix of json and protobuf, could you explain a bit more?

Reply:
ahh, that is the result of copy paste from the AdmissionReview type 😛